### ISSUE TITLE
Orphaned language server spins at 100% CPU after extension host dies (Linux) — mechanism analysis for #178

### ISSUE THREAD

**Issue by NubeBuster** [kisstkondoros/gutter-preview#190]
# Orphaned language server spins at 100% CPU after extension host dies (Linux)

Same symptom as #178 (macOS). This report adds a mechanism analysis from two
live orphans.

## Environment

- Extension: `kisstkondoros.vscode-gutter-preview` 0.32.2 (universal)
- Host: Antigravity IDE (VS Code fork), Electron 39.2.3 / Node 22.21.1, Linux 7.x,
  Wayland/KDE
- Server launched as: `<electron> .../dist/server.js --node-ipc --clientProcessId=<exthost pid>`

## Symptom

Two `dist/server.js` processes whose extension hosts had died kept running,
each pinning one core:

| server | client (ext host) | started | CPU time at capture |
|---|---|---|---|
| A | dead | T+0 | 17 h 18 min |
| B | dead | T−2 days | 17 h 16 min |

RSS ~35 MB, flat — not a memory leak. `utime` ≈ 100 %, `stime` ≈ 0, state `R`,
no syscalls: a pure userland spin.

## Timeline

`dmesg`: `traps: <exthost> trap int3` at T+1m42s (the ext host
died with a V8/Electron fatal). Server A's CPU time equals wall time since
the crash (T+~5 min), i.e. it has spun continuously from the moment its client died. The
second orphan has the same CPU time, so it wedged in the same event.

## Native stack (identical across 3 samples, both PIDs)

```
#14 node::InternalCallbackScope::Close()
#15 node::InternalMakeCallback(...)
#16 node::InternalMakeCallback(...)
#17 node::Environment::CheckImmediate(uv_check_s*)
#18 uv__run_check
#19 uv_run
#20 node::SpinEventLoopInternal
```

Frames 0–13 are unsymbolised JIT code. `InternalCallbackScope::Close()` is
where Node drains the `process.nextTick` / microtask queues after a callback
(here: a `setImmediate` callback — vscode-jsonrpc's `triggerMessageQueue`).
The main thread never leaves that drain: an **unbounded nextTick/microtask
chain**. Timers therefore never run, which is why the extension's own
watchdog never fires:

```js
// dist/server.js:8719 (vscode-languageserver setupExitTimer)
exitTimer = setInterval(() => {
  try { process.kill(processId, 0); }
  catch (ex) { endProtocolConnection(); process.exit(_shutdownReceived ? 0 : 1); }
}, 3e3);
```

The watchdog is the right idea, but it lives on the timer phase that the spin
starves.

## Why the IPC layer does not tear the server down either

`ss -xap` on the `--node-ipc` socketpair: still `ESTAB`, with 5 694 bytes
(server A) and 67 383 bytes (server B) unread in Recv-Q. The server
stopped reading at the crash moment and never received EOF (the peer fd
evidently survived in a child of the dead host), so `disconnect` never fired.

Independently of that, the IPC reader/writer subscribe to an event `process`
never emits:

```js
// dist/server.js:3128, 3143 (vscode-jsonrpc IPCMessageReader/Writer)
eventEmitter.on("close", () => this.fireClose());
eventEmitter.on("close", () => this.fireClose);   // also: not even called
```

Node emits `disconnect` on `process` for a closed IPC channel, not `close`, so
a clean channel close would not close the connection either.

## Suspected origin of the microtask loop (unverified)

The server bundles `node-fetch` + `web-streams-polyfill` for
`ImageCache.store()` of remote images (`_queueMicrotask` pump loops around
`dist/server.js:9742`, `:12283`, `:12388`, `:12451`). A stream pump that
re-queues itself forever once its source breaks matches the observed stack
exactly. A JS-level stack (inspector attach) was not taken.

## Suggested mitigations

1. Make the client-liveness check immune to event-loop starvation — e.g. a
   `worker_threads` watchdog that calls `process.kill(clientPid, 0)` and
   `process.exit()` from the worker (a worker can terminate the process even
   when the main isolate is stuck).
2. Handle `process.on("disconnect")` → `process.exit()` in the server.
3. Bound the stream pump / add a timeout to remote image fetches.

## Unrelated: settings UI shows "Value must be a number"

`package.json` declares `gutterpreview.imagePreviewMaxHeight` /
`imagePreviewMaxWidth` as `"type": "number"` with **string** defaults
`"100"` / `"-1"`; the Settings editor flags them as invalid.

## Zoom lag on every window zoom step (user-visible, likely the main day-to-day symptom)

Changing the window zoom level makes the editor noticeably stutter while this
extension is enabled. The code path:

- `extension.js:24087` rescans on every `onDidChangeTextEditorVisibleRanges`
  (50 ms throttle). A zoom change fires visible-range changes repeatedly for
  every visible editor, so each zoom step triggers a full rescan round trip to
  the server for each of them.
- Each scan result disposes **all** existing decoration types for the document
  (`:24051`) and creates a fresh `createTextEditorDecorationType` per image
  reference per editor (`:23920`). Every decoration type injects its own CSS
  rule with a `gutterIconPath`, so a zoom step tears down and rebuilds every
  gutter-icon rule and forces the renderer to re-decode/re-layout all gutter
  images.

Not a leak (the types are disposed), but the churn scales with visible image
references × visible editors × zoom events. Reusing decoration types keyed by
image path (create once, `setDecorations` with updated ranges) would remove
it.
